proposals/reward-proposal-v2-reward-calculation.md

```diff
@@ -0,0 +1,101 @@
+---
+
+## Problem
+
+Calculating and distributing rewards at the first slot on the epoch boundary is
+not scalable. With increasing number of stake accounts and vote accounts on
+validator nodes, the block time at epoch boundary can last as long as 20 seconds
+because of paying the epoch rewards. Such a long block time degrades the network
+performance and becomes a potential vulnerability of the network.
+
+To solve this problem and speed up the reward time at epoch boundary, an earlier
+approach [Partitioned Inflationary Rewards
+Distribution](https://github.com/solana-labs/solana/blob/master/docs/src/proposals/partitioned-inflationary-rewards-distribution.md)
+is proposed. Let's call this approach `V1`. A prototype of the `V1` proposal is
+implemented. While the `V1` approach works functionally, it, however, has two
+main drawbacks.
+
+First, a restriction is enforced to prevent updates to all stake accounts and
+vote accounts during reward interval. In other words, the stake accounts and
+vote accounts are **locked** during reward interval. All stake account's actions
+such as delegate, redelegate, withdraw etc. are disallowed during reward
+interval and have to wait after reward interval.
+
+Second, there are no on-chain proof for rewards distribution at the epoch. In
+the `V1` approach, the rewards distribution is happening inside bank runtime,
+which update the accounts balance with epoch reward directly. Therefore, there
+is no on-chian records (i.e. transactions) for the rewards distribution. The
+rewards distributions is hidden and happens in a black-box manner.
+
+Given these two drawbacks, we are proposing an enhanced way to handle epoch
+reward - `V2`. The `V2` approach is based on `V1`. However, it completely
+decouples epoch reward into two distinct parts. The first part is to calculate
+the rewards. The second part is to distribute the rewards.
+
+In this proposal, we will focus on the first part - "calculate the reward". A
+follow-up proposal will talk about the second part - "distribute the reward".
+
+## Proposed Solutions
+
+Unlike the current epoch rewards computation approach, in which computing the
+rewards happens at the first slot of epoch boundary in the bank runtime
+directly, the new reward computation is moved to a background service in the
+proposed approach.
+
+This is very similar to the `V1` proposal, a separate service,
+"EpochRewardCalculationService" are created at the start time of the validator.
+This service is listening on a crossbeam channel for incoming rewards
+calculation requests, and handle the requests to compute the rewards.
+
+There are two main difference between the `V2` and `V1` approaches.
+
+The first difference  is that instead of sharing the `StakeCache` with the
+runtime during reward computation. The `StakeCache` is cloned at slot-height 0
+in the `V2` epoch. In this way, we can avoid locking stake accounts and vote
+accounts to prevent updates during the reward calculation.
```

Comment:
Great question. I believe so. After reward interval, the user who has the private key can withdraw the new balance from the account. @t-nelson Can you confirm it?

Reply:
This is a problem. 

> Will it become claimable by anyone who has the private key to the account after it's cleaned up by the runtime?

Yes, however the vast majority of stake account keypairs are discarded. So effectively there's now a new system account with zero data where the stake account previously existed that is bricked.

I feel we need to prevent a stake account from being closed/merged/split while a reward distribution is pending to it. This could be done at the stake program level, with a new syscall (fortunately only needs to be exposed to native programs) that allows it to determine if the distribution is pending

Reply:
> 

If we prevent close/merge/split on the stake account during reward distribution, the end result seems to be the same as what we have proposed V1 - lock the access of stake account? Should just keep V1's behavior here?